src/programs/fastaexplode.1
.TH FASTAEXPLODE 1 "January 2016" BIOBAMBAM2
.SH NAME
fastaexplode - split multi sequence FastA file into one file per sequence
.SH SYNOPSIS
.PP
.B fastaexplode
[options]
.SH DESCRIPTION
fastaexplode reads a FastA file from standard input and writes the contained sequences to a set
of files. It writes one sequence per output file. The file names used are
fasta_<id>.fasta where id is a git digit running decimal number. The prefix
(fasta_ by default) can be set using the -p switch. The .fasta suffix is
omitted if the data only switch is given (see below). The program writes a
two column table to the standard output channel. The first column contains
an output file name and the second the respective sequence id written to
this file.
.PP
The following options can be given
.PP
.B -p<prefix>:
file name prefix. The files will be called <prefix>000000<suffix>, <prefix>000000<suffix><suffix>, etc. where <suffix> is .fasta unless the data only flag is set
.PP
.B -d/--dataonly:
write only the base data but no header/prolog line. This sets the file name suffix to the empty string.
.PP
.B -s/--singleline:
write a single line of base data instead of breaking the base data into multiple lines
.PP
.B -L/--longname:
do not shorten read names by clipping of everything starting from first white space
.PP
.B -l<num>:
set the length of the base data lines (80 by default).
.PP
.B -v/--version:
print the version number and quit
.PP
.B -h/--help:
print a help message and quit
.P
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2016 German Tischler.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.52) unstable; urgency=medium

  * add bamfiltermc program

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 12:39:39 +0200

biobambam2 (2.0.51) unstable; urgency=medium

  * only copy mate cigar string if mate is mapped in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 10:43:26 +0200

biobambam2 (2.0.50) unstable; urgency=medium

  * do not copy CIGAR data of unmapped mates in bamsormadup 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jul 2016 16:13:07 +0200

biobambam2 (2.0.49) unstable; urgency=medium

  * update to libmaus2 version 2.0.216 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 29 Jun 2016 10:38:45 +0200

biobambam2 (2.0.48) unstable; urgency=medium

  * use AdapterFilter object as const in bamadapterfind
  * add blocksortverbose parameter in bamsormadup
  * reflect API update in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 28 Jun 2016 12:28:37 +0200

biobambam2 (2.0.47) unstable; urgency=medium

  * filter out alignments extending past the end of a reference sequence in filtersam

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Jun 2016 12:22:09 +0200

biobambam2 (2.0.46) unstable; urgency=medium

  * add hash sort option in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 12 Jun 2016 14:57:04 +0200

biobambam2 (2.0.45) unstable; urgency=medium

  * do not fail for alignments with no read base data in bamheap2
  * add options for turning N bases randomly to A,C,G,T and to convert bases to upper case in normalisefasta
  * Added a PG line for bam12auxmerge. (by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 10 Jun 2016 11:08:55 +0200

biobambam2 (2.0.44) unstable; urgency=medium

  * reduce input memory per block in bamsormadup if space required is high

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 17 May 2016 10:36:56 +0200

biobambam2 (2.0.43) unstable; urgency=medium

  * improve diagnosis of low memory situations (autoarraytrace option in libmaus2)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 10 May 2016 09:10:42 +0200

biobambam2 (2.0.42) unstable; urgency=medium

  * update to libmaus2 2.0.179 and fix EOF detection bug on pipes 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 02 May 2016 11:03:36 +0200

biobambam2 (2.0.41) unstable; urgency=medium

  * update to libmaus2 2.0.178
  * add number of alignments kept in verbose output of bamfilternames

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 02 May 2016 10:33:25 +0200

biobambam2 (2.0.40) unstable; urgency=medium

  * print memory usage at end of bamcat in verbose mode

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 25 Apr 2016 11:23:05 +0200

biobambam2 (2.0.39) unstable; urgency=medium

  * fix collation in bamtofastq etc concerning comparison of reads names which only differ by leading zeros in number comparison 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 19 Apr 2016 11:25:57 +0200

biobambam2 (2.0.38) unstable; urgency=medium

  * fix for bamrecalculatecigar concerning CIGAR strings starting with a deletion

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Apr 2016 09:27:23 +0200

biobambam2 (2.0.37) unstable; urgency=medium

  * zlib interface changes in libmaus2 for supporting alternative zlibs at run time 

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 07 Apr 2016 13:40:07 +0200

biobambam2 (2.0.36) unstable; urgency=medium

  * add name hash based sub sampling in bamdownsamplerandom

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Apr 2016 08:12:57 +0200

biobambam2 (2.0.35) unstable; urgency=medium

  * use factory class in ranking output for bamtofastq
  * use factory class in bamtoname
  * cleanup bamdownsamplerandom source and use factory class to instantiate alignment file decoder

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Mar 2016 11:12:32 +0200

biobambam2 (2.0.34) unstable; urgency=medium

  * deduce index and md5 filenames from output file name if possible

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Mar 2016 10:19:51 +0200

biobambam2 (2.0.33) unstable; urgency=medium

  * update manual page for bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:30:17 +0100

biobambam2 (2.0.32) unstable; urgency=medium

  * add rmdup option in bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:24:31 +0100

biobambam2 (2.0.31) unstable; urgency=medium

  * fix setting sort order field in bamsort/bamsormadup in case there is no sort order field in the input file 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 07 Mar 2016 15:05:13 +0100

biobambam2 (2.0.30) unstable; urgency=medium

  * do not exit/fail bam indexing when encountering reads with mapping coordinates but flagged as unmapped as sole reads assigned to a reference sequence

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 17 Feb 2016 10:05:51 +0100

biobambam2 (2.0.29) unstable; urgency=medium

  * add murmur3 based digest option in bamseqchksum 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 11 Feb 2016 16:10:33 +0100

biobambam2 (2.0.28) unstable; urgency=medium

  * add bamrecalculatecigar program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 10 Feb 2016 14:14:04 +0100

biobambam2 (2.0.27) unstable; urgency=medium

  * add fastaexplode program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 20 Jan 2016 15:27:26 +0100

biobambam2 (2.0.26) unstable; urgency=medium

  * update to libmaus2 2.0.114
  * Set variable to identify programs to iRODS. (by Andrew Whitwham)

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 12 Jan 2016 19:53:21 +0100

biobambam2 (2.0.25) unstable; urgency=medium

  * update to libmaus2 2.0.102 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 05 Nov 2015 15:28:29 +0100

biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100


src/programs/fastaexplode.cpp
/*
    biobambam2
    Copyright (C) 2016 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <libmaus2/util/ArgParser.hpp>
#include <libmaus2/fastx/StreamFastAReader.hpp>

#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>

static std::string getDefaultPrefix()
{
	return "fasta_";
}

static uint64_t getDefaultLineLength()
{
	return 80;
}

int fastaexplode(libmaus2::util::ArgParser const & arg)
{
	libmaus2::fastx::StreamFastAReaderWrapper SFA(std::cin);
	libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;

	uint64_t cnt = 0;
	std::string const prefixarg =
		arg.uniqueArgPresent("p") ?
			"p" : ( arg.uniqueArgPresent("prefix") ? "prefix" : "" );
	std::string const prefix = prefixarg.size() ? arg[prefixarg] : getDefaultPrefix();
	bool const singleline = arg.argPresent("s") || arg.argPresent("singleline");
	bool const longname = arg.argPresent("L") || arg.argPresent("longname");
	bool const dataonly = arg.argPresent("d") || arg.argPresent("dataonly");
	uint64_t const linelength = arg.uniqueArgPresent("c") ? arg.getUnsignedNumericArg<uint64_t>("c") : getDefaultLineLength();

	while ( SFA.getNextPatternUnlocked(pattern) )
	{
		std::ostringstream fnostr;
		fnostr << prefix << std::setw(6) << std::setfill('0') << cnt++ << std::setw(0) << (dataonly ? "" : ".fasta");
		std::string const fn = fnostr.str();
		libmaus2::aio::OutputStreamInstance OSI(fn);

		std::string & spat = pattern.spattern;

		for ( uint64_t i = 0; i < spat.size(); ++i )
			spat[i] = toupper(spat[i]);

		if ( !longname )
			pattern.sid = pattern.getShortStringId();

		if ( dataonly )
			OSI.write(pattern.spattern.c_str(),pattern.spattern.size());
		else if ( singleline )
			OSI << pattern;
		else
			pattern.printMultiLine(OSI,linelength);

		std::cout << fn << "\t" << pattern.getStringId() << std::endl;
	}

	return EXIT_SUCCESS;
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::util::ArgParser const arg(argc,argv);

		if (
			arg.uniqueArgPresent("v") || arg.uniqueArgPresent("version")
		)
		{
			std::cerr << ::biobambam2::Licensing::license();
			return EXIT_SUCCESS;
		}
		else if (
			arg.uniqueArgPresent("h") || arg.uniqueArgPresent("help")
		)
		{
			std::cerr << ::biobambam2::Licensing::license();
			std::cerr << std::endl;
			std::cerr << "options:" << std::endl;
			std::cerr << std::endl;

			std::vector< std::pair<std::string,std::string> > V;

			V.push_back ( std::pair<std::string,std::string> ( "-v/--version", "print version number and quit" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-h/--help", "print help message and quit" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-s/--singleline", "do not wrap sequence data lines" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-L/--longname", "do not shorten name" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-l<cols>", "line length (default: "+libmaus2::util::NumberSerialisation::formatNumber(getDefaultLineLength(),0)+")" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-d/--dataonly", "do not print FastA header (data only)" ) );
			V.push_back ( std::pair<std::string,std::string> ( "-p", std::string("prefix for FastA output files (default: ")+getDefaultPrefix()+")" ) );

			::biobambam2::Licensing::printMap(std::cerr,V);

			std::cerr << std::endl;
			return EXIT_SUCCESS;

		}

		return fastaexplode(arg);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
